tools/src/main/kotlin/org/dashevo/tools/DashPayWallet.kt

```diff
@@ -0,0 +1,291 @@
+                        val contactIdentity = platform.identities.get(contactRequest.toUserId.toBase58())
+                        if (encryptionKey == null && wallet.isEncrypted) {
+                            // Don't bother with DeriveKeyTask here, just call deriveKey
+                            encryptionKey = wallet.keyCrypter!!.deriveKey(password)
+                        }
+                        blockchainIdentity.addPaymentKeyChainFromContact(contactIdentity!!, it, encryptionKey)
+                        addedContact = true
+                    }
+                } catch (e: KeyCrypterException) {
+                    // we can't send payments to this contact due to an invalid encryptedPublicKey
+                    log.info("ContactRequest: error ${e.message}")
+                }
+            }
+            // Get all contact requests where toUserId == userId, the users who have added me
+            val fromContactDocuments = ContactRequests(platform).get(userId, toUserId = true, afterTime = lastContactRequestTime, retrieveAll = true)
+            fromContactDocuments.forEach {
+                val contactRequest = ContactRequest(it)
+                userIdList.add(contactRequest.ownerId)
+                contactRequests.add(contactRequest)
+
+                // add the sending to contact keychain if it doesn't exist
+                val contact = EvolutionContact(userId, contactRequest.ownerId)
+                try {
+                    if (!wallet.hasSendingKeyChain(contact)) {
+                        val contactIdentity = platform.identities.get(contactRequest.ownerId)
+                        if (encryptionKey == null && wallet.isEncrypted) {
+                            encryptionKey = wallet.keyCrypter!!.deriveKey(password)
+                        }
+                        blockchainIdentity.addContactPaymentKeyChain(contactIdentity!!, it, encryptionKey)
+                        addedContact = true
+                    }
+                } catch (e: KeyCrypterException) {
+                    // we can't send payments to this contact due to an invalid encryptedPublicKey
+                    log.info("ContactRequest: error ${e.message}")
+                }
+            }
+
+            // If new keychains were added to the wallet, then update the bloom filters
+            if (addedContact) {
+                peerGroup?.recalculateFastCatchupAndFilter(PeerGroup.FilterRecalculateMode.SEND_IF_CHANGED)
+            }
+
+            //obtain profiles from new contacts
+            if (userIdList.isNotEmpty()) {
+                updateContactProfiles(userIdList.toList(), 0L)
+            }
+
+            // fetch updated profiles from the network
+            updateContactProfiles(userId, lastContactRequestTime)
+
+            // fire listeners if there were new contacts
+            if (fromContactDocuments.isNotEmpty() || toContactDocuments.isNotEmpty()) {
+            //    fireContactsUpdatedListeners()
+            }
```

Comment:
There may be no need for this section.  Or perhaps there is if we later add listeners or something.  in `dash-wallet` there is LiveData that is listening here!